File: SS/Labs/Lab1/1_regex_to_dfa.c
#include <stdio.h>
#include <string.h>

#define MAX 128

typedef struct {
    int from;
    char sym;
    int to;
} Transition;

Transition T[MAX];
int tcount = 0;
int final_state = 0;

void add_tr(int from, char sym, int to) {
    T[tcount].from = from;
    T[tcount].sym = sym;
    T[tcount].to = to;
    tcount++;
}

void build_dfa(const char *re) {
    int cur = 0;
    tcount = 0;
    int n = (int)strlen(re);

    for (int i = 0; i < n; ++i) {
        char ch = re[i];
        char nxt = (i + 1 < n) ? re[i + 1] : '\0';

        if (ch == '*' || ch == '+' || ch == '|') continue;

        if (nxt == '*') {
            add_tr(cur, ch, cur);
            i++;
        } else if (nxt == '+') {
            add_tr(cur, ch, cur + 1);
            add_tr(cur + 1, ch, cur + 1);
            cur += 1;
            i++;
        } else {
            add_tr(cur, ch, cur + 1);
            cur += 1;
        }
    }
    final_state = cur;
}

int validate(const char *s) {
    int state = 0;
    for (int i = 0; s[i]; ++i) {
        char ch = s[i];
        int moved = 0;
        for (int j = 0; j < tcount; ++j) {
            if (T[j].from == state && T[j].sym == ch) {
                state = T[j].to;
                moved = 1;
                break;
            }
        }
        if (!moved) return 0;
    }
    return state == final_state;
}

void print_dfa(void) {
    printf("\nDFA\n");
    printf("%-8s %-8s %-8s\n", "State", "Input", "Next");
    printf("------------------------\n");
    for (int i = 0; i < tcount; ++i) {
        printf("q%-7d %-8c q%-7d\n", T[i].from, T[i].sym, T[i].to);
    }
    printf("------------------------\n");
    printf("Start: q0\nFinal: q%d\n", final_state);
}

int main() {
    char re[64], s[64], c;
    printf("Enter regex: ");
    scanf("%63s", re);

    build_dfa(re);
    print_dfa();

    do {
        printf("\nTest a string? (y/n): ");
        scanf(" %c", &c);
        if (c == 'y' || c == 'Y') {
            printf("String: ");
            scanf("%63s", s);
            if (validate(s)) printf("ACCEPTED\n");
            else printf("REJECTED\n");
        }
    } while (c == 'y' || c == 'Y');
    return 0;
}

File: SS/Labs/PythonLabs/Lab1/1_regex_to_dfa.py
# Lab 1 — Regex to DFA (linear regex: literals, *, +)

transitions = []
final_state = 0

def build_dfa(re):
    global transitions, final_state
    transitions = []
    cur = 0
    i = 0
    while i < len(re):
        ch = re[i]
        nxt = re[i + 1] if i + 1 < len(re) else '\0'

        if ch in ('*', '+', '|'):
            i += 1
            continue

        if nxt == '*':
            transitions.append((cur, ch, cur))
            i += 2
        elif nxt == '+':
            transitions.append((cur, ch, cur + 1))
            transitions.append((cur + 1, ch, cur + 1))
            cur += 1
            i += 2
        else:
            transitions.append((cur, ch, cur + 1))
            cur += 1
            i += 1

    final_state = cur

def validate(s):
    state = 0
    for ch in s:
        moved = False
        for (frm, sym, to) in transitions:
            if frm == state and sym == ch:
                state = to
                moved = True
                break
        if not moved:
            return False
    return state == final_state

def print_dfa():
    print("\nDFA")
    print(f"{'State':<8} {'Input':<8} {'Next':<8}")
    print("-" * 24)
    for (frm, sym, to) in transitions:
        print(f"q{frm:<7} {sym:<8} q{to:<7}")
    print("-" * 24)
    print(f"Start: q0\nFinal: q{final_state}")

def main():
    re = input("Enter regex: ").strip()
    build_dfa(re)
    print_dfa()

    while True:
        c = input("\nTest a string? (y/n): ").strip().lower()
        if c == 'y':
            s = input("String: ").strip()
            print("ACCEPTED" if validate(s) else "REJECTED")
        else:
            break

if __name__ == "__main__":
    main()
